hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-replication-core/src/test/java/com/google/cloud/bigtable/hbase/replication/CloudBigtableReplicatorTest.java

```diff
@@ -437,4 +444,275 @@ public void testReplicateFailsToSubmitTask() throws IOException {
+    replicator.start(sharedResources, incompatibleMutationAdapter, 2000, false);
+
+    // Enable two-way replication manually.
+    Configuration replicationConf = new Configuration(false);
+    replicationConf.setBoolean(ENABLE_TWO_WAY_REPLICATION_MODE_KEY, true);
+    replicator.maybeStartTwoWayReplication(replicationConf, mockMetricExporter);
+
+    // Row key 1 WAL comes from CBT and should be filtered out.
+    Cell cell11 = new KeyValue(getRowKey(1), CF1, null, TIMESTAMP, getValue(1));
+    // Special replicated mutation from CBT.
+    Cell cell1s =
+        new KeyValue(
+            getRowKey(1), CF1, DEFAULT_SOURCE_CBT_QUALIFIER.getBytes(), 0l, KeyValue.Type.Delete);
+
+    // Row key 2 WAL comes from HBase and should be replicated.
+    Cell cell21 = new KeyValue(getRowKey(2), CF1, null, TIMESTAMP, getValue(1));
+    Cell cell2s =
+        new KeyValue(
+            getRowKey(2), CF1, DEFAULT_SOURCE_HBASE_QUALIFIER.getBytes(), 0l, KeyValue.Type.Delete);
+
+    // Create respective WAL entries.
+    // Entry 1 will be filtered out.
+    BigtableWALEntry walEntry1 =
+        new BigtableWALEntry(TIMESTAMP, Arrays.asList(cell11, cell1s), TABLE_NAME_STRING);
+    // Entry 2 should be replicated.
+    BigtableWALEntry walEntry2 =
+        new BigtableWALEntry(TIMESTAMP, Arrays.asList(cell21), TABLE_NAME_STRING);
+    BigtableWALEntry walEntry2s =
+        new BigtableWALEntry(TIMESTAMP, Arrays.asList(cell21, cell2s), TABLE_NAME_STRING);
+
+    // Create WAL for replicator.
+    Map<String, List<BigtableWALEntry>> walsToReplicate = new HashMap<>();
+    walsToReplicate.put(TABLE_NAME_STRING, Arrays.asList(walEntry1, walEntry2));
+
+    // Create expected state of replicationTask input.
+    Map<ByteRange, List<Cell>> expectedBatchOfWal = new HashMap<>();
+    expectedBatchOfWal.put(new SimpleByteRange(getRowKey(2)), walEntry2s.getCells());
+
+    // Everything succeeds.
+    when(mockExecutorService.submit((Callable<Object>) any()))
+        .thenReturn(CompletableFuture.completedFuture(true));
+
+    // Trigger replication
+    assertTrue(replicator.replicate(walsToReplicate));
+
+    // called during constructor
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(DROPPED_INCOMPATIBLE_MUTATION_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_DELETES_METRICS_KEY, 0);
+    verify(mockMetricExporter).incCounters(INCOMPATIBLE_MUTATION_TIMESTAMP_OVERFLOW_METRIC_KEY, 0);
+    verify(mockMetricExporter).incCounters(PUTS_IN_FUTURE_METRIC_KEY, 0);
+    // Metrics reflect that one WAL entry is dropped and one passes through.
+    verify(mockMetricExporter).incCounters(SOURCE_CBT_DROPPED, 1);
+    verify(mockMetricExporter).incCounters(SOURCE_HBASE_REPLICATED, 1);
+
+    verify(mockExecutorService)
+        .submit(
+            new CloudBigtableReplicationTask(
+                TABLE_NAME_STRING, mockConnection, expectedBatchOfWal));
```

Comment:
double check that this is matched, drop a cell from expectedBatch and check if test fails. I think this is fine, but pl double check.

Reply:
Double checked. Removing either the sourceTagged cell or normal cell will fail this call.